### PR TITLE
parallelize tests on circleci using workflows

### PR DESCRIPTION
**Pull request by wwymak**
parallelize tests on circleci using [workflows](https://circleci.com/docs/2.0/workflows/) 
- also fixed the dependency caching
- have to comment out the test_full_transformer as it's causing timeouts on ci
- at the moment it doesn't seem to be reporting final test result correctly but this might be something that needs to be set on github https://circleci.com/docs/2.0/workflows-waiting-status/

### CODE REVIEW

File: flood_forecast/transformer_xl/dummy_torch.py [hunk 1]
```diff
@@ -13,5 +13,5 @@ def __init__(self, forecast_length:int):
 
     def forward(self, x:torch.Tensor, mask=None):
         batch_sz = x.size(0)
-        result = torch.ones(batch_sz, self.out_len, requires_grad=True)
+        result = torch.ones(batch_sz, self.out_len, requires_grad=True, device=x.device)
```
Comment (isaacmg) [inline flood_forecast/transformer_xl/dummy_torch.py:16]: Was this for CUDA compatibility?

File: tests/pytorc_train_tests.py [hunk 2]
```diff
@@ -12,34 +11,58 @@
...
         self.opt = torch.optim.Adam(self.dummy_model.model.parameters(), lr=0.0001)
         self.criterion = torch.nn.modules.loss.MSELoss()
         self.data_loader = DataLoader(self.dummy_model.training, batch_size=2, shuffle=False, sampler=None,
-           batch_sampler=None, num_workers=0, collate_fn=None,
-           pin_memory=False, drop_last=False, timeout=0,
-           worker_init_fn=None)
-        # TODO add LSTM test self.lstm = 
+                                      batch_sampler=None, num_workers=0, collate_fn=None,
+                                      pin_memory=False, drop_last=False, timeout=0,
+                                      worker_init_fn=None)
```
Comment (isaacmg) [inline tests/pytorc_train_tests.py:64]: Did anything change here?

File: flood_forecast/transformer_xl/dummy_torch.py (hunk 1 above)
Comment (wwymak) [reply flood_forecast/transformer_xl/dummy_torch.py:16]: yeap-- when I was running it locally the tensors are not in the same device as the model

File: tests/pytorc_train_tests.py (hunk 2 above)
Comment (wwymak) [reply tests/pytorc_train_tests.py:64]: no, just white spaces (it's the indent thing in pep8)

**Comment by wwymak** [comment]
> Also do you know why ci/circleci:build is still listed as expected instead of passing? 

I think it has to do with a setting as per: https://circleci.com/docs/2.0/workflows-waiting-status/ (but I don't have the admin rights to check the settings on the repo-- can you see if the settings need to be tweaked @isaacmg ?)

> If you think it is fine for now though you can merge this then you can address that in a separate PR

I think it might be easier to figure out how to fix this in a separate PR as per #80  ?

**Comment by Pranjalya** [comment]
Can you set the unbuffered environment variable, as per [here](https://support.circleci.com/hc/en-us/articles/360045268074-Build-Fails-with-Too-long-with-no-output-exceeded-10m0s-context-deadline-exceeded-). This may resolve timeouts.

**Comment by wwymak** [comment]
> Can you set the unbuffered environment variable, as per [here](https://support.circleci.com/hc/en-us/articles/360045268074-Build-Fails-with-Too-long-with-no-output-exceeded-10m0s-context-deadline-exceeded-). This may resolve timeouts.

yeah, I might try that as well for #80, but the main issue to fix is the test taking so long since it wastes circleci credits and also it's affects how productive you are if you have to wait half an hour for some simple tests to finish...

**Pull request merged**